### PR TITLE
Fix wrong path encoding

### PR DESCRIPTION
**Pull request by ianton-ru**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Solved #1536
Fix incorrect encoding special characters in S3 paths.

### Documentation entry for user-facing changes
...

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [x]  All with TSAN
- [x]  All with MSAN
- [x]  All with UBSAN
- [x]  All with Coverage
- [ ]  All with Aarch64
- [x]  All Regression
- [ ]  Disable CI Cache

#### Regression jobs to run:
- [ ]  Fast suites (mostly <1h)
- [ ]  Aggregate Functions (2h)
- [ ]  Alter (1.5h)
- [ ]  Benchmark (30m)
- [ ]  ClickHouse Keeper (1h)
- [ ]  Iceberg (2h)
- [ ]  LDAP (1h)
- [ ]  Parquet (1.5h)
- [ ]  RBAC (1.5h)
- [ ]  SSL Server (1h)
- [ ]  S3 (2h)
- [ ]  Tiered Storage (2h)

### CODE REVIEW

**Comment by ianton-ru** [comment]
AI audit note: This review comment was generated by AI (`gpt-5.4`).

Audit update for PR #1537 after the latest branch update:

## Confirmed defects

No confirmed defects in reviewed scope.

## Coverage summary

Scope reviewed: the updated `DB::S3::URI::URI` flow in `src/IO/S3/URI.cpp`, especially the two remaining changed paths `Poco::URI(uri_str, false)` and `URIConverter::modifyURI(..., false)`, plus the Iceberg/object-storage call chain that feeds absolute S3 paths into `S3::URI`, existing S3 URI unit coverage in `src/IO/tests/gtest_s3_uri.cpp`, the new Iceberg integration coverage in `tests/integration/test_database_iceberg/test.py`, and adjacent scheme-mapper/archive-syntax coverage.
Categories failed: none.
Categories passed: raw `%xx` object-key preservation on initial parse; scheme-mapper remap preserving encoded path/query bytes; compatibility of the unchanged non-`versionId` `?` fallback with existing S3 URI tests; archive-syntax split occurring before the changed parse step; no confirmed concurrency/lifetime/rollback issues in scope.
Assumptions/limits: static audit only, no runtime execution; there is still limited direct test coverage for combinations of encoded paths with scheme-mapper plus archive/glob edge cases, but I did not confirm a defect from the latest changed lines.

**Review by mkmkme — DISMISSED** [review]

**Comment by subkanthi** [comment]
@ianton-ru , u are probably addressing this one too https://github.com/Altinity/ClickHouse/pull/1516 as part of this PR.

**Comment by ianton-ru** [comment]
Test `03212_variant_dynamic_cast_or_default` is flaky, see https://github.com/clickhouse/clickhouse/issues/87583

**Comment by ianton-ru** [comment]
Issue was introduced in #1092.
In Iceberg metadata paths are already in correct form
S3::URI decodes path, and column=12%3A00%3A00  converted into column=12:00:00 . Before #1092 paths are used "as is".
It's impossible to encode it back later, because character =  also encoded into %3D in this case. You can make string as partition key, and put in string =  character, in this case path looks like column=foo%3Dbar , after decoding it is converted into column=foo=bar  and you lost information, which character was encoded and which wasn't.

**Review by mkmkme — DISMISSED** [review]

**Comment by mkmkme** [comment]
@codex review

**Comment by ianton-ru** [comment]
> @ianton-ru , u are probably addressing this one too #1516 as part of this PR.

@subkanthi Looks like your solution gives the same result, and more elegant.

File: src/IO/S3/URI.cpp [hunk 1]
```diff
@@ -54,7 +55,7 @@ URI::URI(const std::string & uri_, bool allow_archive_path_syntax)
     else
         uri_str = uri_;
 
-    uri = Poco::URI(uri_str);
+    uri = Poco::URI(uri_str, enable_url_encoding);
```
[the remark this answers is not in the document]
Comment (ianton-ru) [reply src/IO/S3/URI.cpp:58]: This methos is called not with any URL, but only with path in S3 or Azure storage. Question mark should be encoded in S3 object name https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html as well as in Azure https://learn.microsoft.com/en-us/azure/devops/organizations/settings/naming-restrictions?view=azure-devops&tabs=git.

**Comment by ianton-ru** [comment]
@codex review

**Comment by alsugiliazova** [comment]
Regression run https://github.com/Altinity/clickhouse-regression/actions/runs/23306333458

**Comment by alsugiliazova** [comment]
# PR #1537 CI Verification Report

## CI Results: 54 success, 1 failure, 109 skipped

### Failures (1)

| # | Suite | Error | Related to PR |
|---|-------|-------|---------------|
| 1 | GrypeScanServer (-alpine) / Grype Scan | CVE scan | **No** — container image security scan |

### PR's Own Integration Tests

- **Integration tests (amd_asan, flaky check):** **Passed** — CI detected changes in `test_database_iceberg/test.py` and ran the flaky check (5 repetitions). All passed, including the new `test_partitioning_by_time` and `test_partitioning_by_string` tests.

### Passed Suites

- **Builds:** amd_debug, amd_asan, amd_binary, amd_release, arm_binary, arm_release — all passed
- **Unit tests:** asan — passed
- **Stateless tests (14 suites):** All passed (amd_asan ×3, amd_binary ×4, amd_debug ×6, arm_binary ×2)
- **Integration tests (15 shards):** All passed (amd_asan old analyzer ×6, amd_binary ×5, arm_binary distributed plan ×4)
- **Stress test:** amd_debug — passed
- **AST fuzzer:** amd_debug — passed
- **BuzzHouse:** amd_debug — passed

### Skipped Suites (109)

Excluded by PR author: all TSAN, MSAN, UBSAN, Coverage, and Regression suites.

## Conclusion

**PR #1537 CI is green.**
